src/lib/list.h

```diff
@@ -0,0 +1,218 @@
+/*
+ * Copyright (c) 2017, [Ribose Inc](https://www.ribose.com).
+ * All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without modification,
+ * are permitted provided that the following conditions are met:
+ *
+ * 1.  Redistributions of source code must retain the above copyright notice,
+ *     this list of conditions and the following disclaimer.
+ *
+ * 2.  Redistributions in binary form must reproduce the above copyright notice,
+ *     this list of conditions and the following disclaimer in the documentation
+ *     and/or other materials provided with the distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND
+ * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
+ * WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+ * DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE
+ * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+ * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+ * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
+ * OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF
+ * THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+/** Doubly linked list
+ *  @file
+ */
+#ifndef RNP_LIST_H
+#define RNP_LIST_H
+
+#include <stddef.h>
+
+/**
+ *  @private
+ *  Note that all list_item pointers returned from the
+ *  API are actually pointers directly to your data,
+ *  and thus can be directly cast to the appropriate
+ *  type.
+ *
+ *  The code below will print:
+ *      first
+ *      second
+ *
+ *  @code
+ *  // always initialize to NULL
+ *  list lst = NULL;
+ *  list_append(&lst, "second", strlen("second") + 1);
+ *  list_insert(&lst, "first", strlen("first") + 1);
+ *  assert(list_length(lst) == 2);
+ *
+ *  list_item *item = list_front(lst);
+ *  while (item) {
+ *      printf("%s\n", (char*)item);
+ *      item = list_next(item);
+ *  }
+ *  @endcode
+ *
+ *  Searching:
+ *  The code below searches for the value 1 and will
+ *  output:
+ *      Found 1
+ *      Found 1
+ *      Found 1
+ *
+ *  @code
+ *  list lst = NULL;
+ *  static const int some_ints[] = {0, 1, 1, 2, 3, 4, 1};
+ *  for (int i = 0; i < sizeof(some_ints) / sizeof(some_ints[0]); i++) {
+ *      list_append(&lst, &some_ints[i], sizeof(some_ints[i]));
+ *  }
+ *
+ *  int one = 1;
+ *  list_item *item = list_find(lst, &one, sizeof(one));
+ *  while (item) {
+ *      printf("Found %d\n", *(int*)item);
+ *      item = list_find_next(item, &one, sizeof(one));
+ *  }
+ *  @endcode
+ */
+
+typedef struct list_head *list;
+typedef struct list_head  list_head;
+typedef struct list_item  list_item;
```

Comment:
maybe use `rnp_` suffix?

Also, `typedef struct list_head *list` may confuse and I think better to use `typedef struct list_head list`.